src/renderer/backend/dx9.rs

```diff
@@ -384,8 +396,32 @@ impl TextureHeap {
             )
         })?;
 
+        let id = TextureId::from(self.textures.len());
+        self.textures.push(Texture { resource, id, width, height });
+
+        Ok(id)
+    }
+
+    unsafe fn upload_texture(
+        &mut self,
+        texture_id: TextureId,
+        data: &[u8],
+        width: u32,
+        height: u32,
+    ) -> Result<()> {
+        let texture = &self.textures[texture_id.id()];
+        if texture.width != width || texture.height != height {
+            return Err(Error::new(
```

Comment:
Same observation as above, just marking the spot with a  comment for convenience.